geniza/corpus/models.py

```diff
@@ -434,28 +451,78 @@ def editions(self):
     @classmethod
     def items_to_index(cls):
         """Custom logic for finding items to be indexed when indexing in
-        bulk; only include public docs."""
-        return cls.objects.filter(status=Document.PUBLIC)
+        bulk."""
+        # TODO: can we share common/reused prefetching logic
+        # in a custom qureyset filter or similar? (adapted here from admin)
+        return cls.objects.select_related("doctype").prefetch_related(
+            "tags",
+            "languages",
+            "footnotes",
+            "log_entries",
+            Prefetch(
+                "textblock_set",
+                queryset=TextBlock.objects.select_related(
+                    "fragment", "fragment__collection"
+                ),
+            ),
+        )
 
     def index_data(self):
         """data for indexing in Solr"""
-
         index_data = super().index_data()
+
+        # get fragments via textblocks for correct order
+        # and to take advantage of prefetching
+        fragments = [tb.fragment for tb in self.textblock_set.all()]
         index_data.update(
             {
                 "pgpid_i": self.id,
                 "type_s": self.doctype.name if self.doctype else "Unknown",
                 "description_t": self.description,
-                "notes_t": self.notes,
-                "needs_review_t": self.needs_review,
-                "shelfmark_ss": [f.shelfmark for f in self.fragments.all()],
+                "notes_t": self.notes or None,
+                "needs_review_t": self.needs_review or None,
+                "shelfmark_ss": [f.shelfmark for f in fragments],
+                # library/collection possibly redundant?
+                "collection_ss": [str(f.collection) for f in fragments],
                 "tags_ss": [t.name for t in self.tags.all()],
                 "status_s": self.get_status_display(),
                 "old_pgpids_is": self.old_pgpids,
-                # TODO: editors/translators/sources
             }
         )
 
+        # count scholarship records by type
+        footnotes = self.footnotes.all()
+        counts = defaultdict(int)
+        for fn in footnotes:
+            for val in fn.doc_relation:
+                counts[val] += 1
```

Comment:
wonder if this thing could also become a method that uses annotation/aggregation to accomplish this faster, although maybe we won't really have enough footnotes of each type to make it worthwhile

Reply:
hmm, interesting — we can probably generate counts in the database, but that would complicate the case when we index a single record